xmodule/library_content_block.py

```diff
@@ -251,6 +234,64 @@ def _publish_event(self, event_name, result, **kwargs):
         self.runtime.publish(self, f"edx.librarycontentblock.content.{event_name}", event_data)
         self._last_event_result_count = len(result)  # pylint: disable=attribute-defined-outside-init
 
+    @classmethod
+    def _get_valid_children(cls, library_children, candidates, manual):
+        """
+        Dynamically selects block_ids which children are possible for selection
+        """
+        if candidates and manual:
+            return candidates
+        return {(child.block_type, child.block_id) for child in library_children}
+
+    @classmethod
+    def make_selection(cls, selected, library_children, candidates, max_count, manual, shuffle):
```

Comment:
This is already a complex method (perhaps the most complex and important method in the whole project), but the mixing of sets and lists, as well as the mixing of `.difference/.intersection` vs. `-/&`, is making it extra hard to follow.

Please:
* Convert to sets, unless order matters.
* Use type annotations (e.g., `set[tuple[str, str]]` and `list[tuple[str, str]]`)
* Use set operators (`&`, `-`, and `|`) rather than set methods (`intersection`, `difference`, and `union`).

Reply:
What I decided here is that list might actually be the more useful data type for the for the following reasons:

We need to preserve ordering of the components across selections, and sets are unordered.
We need to do list operations many more times than set operations. Specifically, we need to truncate the selections by length and concatenate lists while preserving ordering. 

In summation, Although I am changing details, the ethos of your comment is being applied: make things consistent.

Reply:
Thanks for switching from set methods to set operators; those do help.

> What I decided here is that list might actually be the more useful data type for the for the following reasons:

I wouldn't say that either lists or sets are more useful in this function; sorry if it seemed like I was nudging you to pick one or the other. They both have a place in this function. Converting everything to a list does not make the function more readable or correct. Conversely, it obscures which variables have a defined order.

There's a bug in new implementation that I want to use as an example. Look at valid_old_block_keys. Semantically, this should be a set, since its order is arbitrary (that is: neither well-defined nor truly random), but you're converting it to a list and then slicing that list in order to decide which keys to boot out:

```python
new_selected = valid_old_block_keys[0:size]
overlimit_block_keys = valid_old_block_keys[size:]
```

This will mean that the blocks will not get booted out randomly, as your docstring helpfully says they should be; rather, they'll be booted out at the whim of the order of valid_old_block_keys, which could be biased towards the blocks at the end of the `children` list, or perhaps the alphabetical order of the block keys, or something else (Python make no guarantees).

In the old function, this was done using sets, which meant they couldn't be sliced:
```python
overlimit_block_keys = set(rand.sample(selected_keys, num_to_remove))
selected_keys -= overlimit_block_keys
```

I realize that the logic of the new function is necessarily more complex (due to candidates being a thing now), but that's all the more reason to be careful with ordering.

Also, please please please add type annotations.

`old_selected: set = {tuple(k) for k in selected}`

vs

`old_selected: list = [tuple(k) for k in selected]`.